staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,239 @@ limitations under the License.
+	apiServices := make([]*v1.APIService, 0, len(s.specsByAPIServiceName))
+	for k := range s.specsByAPIServiceName {
+		apiServices = append(apiServices, &s.specsByAPIServiceName[k].apiService)
+	}
+	sortByPriority(apiServices)
+	caches := make([]cached.Data[*spec.Swagger], len(s.specsByAPIServiceName))
+	for i, apiService := range apiServices {
+		caches[i] = &(s.specsByAPIServiceName[apiService.Name].spec)
+	}
+
+	return cached.NewListMerger(func(results []cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		var merged *spec.Swagger
+		etags := make([]string, 0, len(results))
+		for _, specInfo := range results {
+			result := specInfo.Get()
+			if result.Err != nil {
+				// APIService name and err message will be included in
+				// the error message as part of decorateError
+				klog.Warning(result.Err)
+				continue
+			}
+			if merged == nil {
+				merged = &spec.Swagger{}
+				*merged = *result.Data
+				merged.Paths = nil
+				merged.Definitions = nil
+			}
+			etags = append(etags, result.Etag)
+			if err := aggregator.MergeSpecsIgnorePathConflict(merged, result.Data); err != nil {
+				return cached.NewResultErr[*spec.Swagger](fmt.Errorf("failed to build merge specs: %v", err))
+			}
 		}
-		return err
-	}
-	return nil
+		// Printing the etags list is stable because it is sorted.
+		return cached.NewResultOK(merged, fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%#v", etags)))))
+	}, caches)
 }
 
-// tryDeleteServiceSpecs tries delete specified specInfo from openAPISpecs map, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryDeleteServiceSpecs(apiServiceName string) error {
-	orgSpecInfo, exists := s.openAPISpecs[apiServiceName]
+// updateServiceLocked updates the spec cache by downloading the latest
+// version of the spec.
+func (s *specAggregator) updateServiceLocked(name string) error {
+	specInfo, exists := s.specsByAPIServiceName[name]
 	if !exists {
-		return nil
-	}
-	delete(s.openAPISpecs, apiServiceName)
-	klog.V(2).Infof("Updating OpenAPI spec because %s is removed", apiServiceName)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		regenerationCounter.With(map[string]string{"apiservice": apiServiceName, "reason": "delete"})
-		regenerationDurationGauge.With(map[string]string{"reason": "delete"}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		s.openAPISpecs[apiServiceName] = orgSpecInfo
-		return err
+		return ErrAPIServiceNotFound
 	}
-	return nil
+	result := specInfo.downloader.Get()
+	filteredResult := cached.NewTransformer[*spec.Swagger](func(result cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		if result.Err != nil {
+			return result
+		}
+		return cached.NewResultOK(aggregator.FilterSpecByPathsWithoutSideEffects(result.Data, []string{"/apis/"}), result.Etag)
+	}, result)
+	specInfo.spec.Replace(filteredResult)
```

Comment:
I don't think that's super bad, but since we're recreating the transformer every minute here, we're invalidating at least that step of cache, right?

Reply:
Yes, but since the returned etag will be the same (if no change), rest of the steps are not executed. Could be optimized with maybe updates to the caching code but I see it as much less of a problem

Reply:
Agreed, but I think it shows that the framework has a limitation that doesn't let you do the right thing easily. I think in that specific case it's fine, but if it had not, you would have had to write more code (like you did in the other PR).